src/coreComponents/physicsSolvers/multiphysics/MultiphasePoroelasticKernel.hpp

```diff
@@ -0,0 +1,537 @@
+  real64 complete( localIndex const k,
+                   StackVariables & stack ) const
+  {
+    GEOSX_UNUSED_VAR( k );
+    real64 maxForce = 0;
+
+    CONSTITUTIVE_TYPE::KernelWrapper::DiscretizationOps::template fillLowerBTDB< numNodesPerElem >( stack.localJacobian );
+
+    //int nFlowDof = m_numComponents + 1;
+
+    constexpr int nUDof = numNodesPerElem * numDofPerTestSupportPoint;
+
+    for( int localNode = 0; localNode < numNodesPerElem; ++localNode )
+    {
+      for( int dim = 0; dim < numDofPerTestSupportPoint; ++dim )
+      {
+        localIndex const dof = LvArray::integerConversion< localIndex >( stack.localRowDofIndex[numDofPerTestSupportPoint * localNode + dim] - m_dofRankOffset );
+        if( dof < 0 || dof >= m_matrix.numRows() ) continue;
+        m_matrix.template addToRowBinarySearchUnsorted< parallelDeviceAtomic >( dof,
+                                                                                stack.localRowDofIndex,
+                                                                                stack.localJacobian[numDofPerTestSupportPoint * localNode + dim],
+                                                                                numNodesPerElem * numDofPerTrialSupportPoint );
+
+        RAJA::atomicAdd< parallelDeviceAtomic >( &m_rhs[dof], stack.localResidual[numDofPerTestSupportPoint * localNode + dim] );
+        maxForce = fmax( maxForce, fabs( stack.localResidual[numDofPerTestSupportPoint * localNode + dim] ) );
+
+        m_matrix.template addToRowBinarySearchUnsorted< parallelDeviceAtomic >( dof,
+                                                                                stack.localFlowDofIndex,
+                                                                                stack.localDispFlowJacobian[numDofPerTestSupportPoint * localNode + dim],
+                                                                                m_numComponents + 1 );
+
+      }
+    }
+
+    for( localIndex i = 0; i < m_numComponents; ++i )
+    {
+      localIndex const dof = LvArray::integerConversion< localIndex >( stack.localFlowDofIndex[i] - m_dofRankOffset );
+      if( dof < 0 || dof >= m_matrix.numRows() )
+        continue;
+      m_matrix.template addToRowBinarySearchUnsorted< parallelDeviceAtomic >( dof,
+                                                                              stack.localRowDofIndex,
+                                                                              stack.localFlowDispJacobian[i],
+                                                                              nUDof );
+      m_matrix.template addToRowBinarySearchUnsorted< parallelDeviceAtomic >( dof,
+                                                                              stack.localFlowDofIndex,
+                                                                              stack.localFlowFlowJacobian[i],
+                                                                              m_numComponents + 1 );
```

Comment:
Use `addToRow< serialAtomic >` instead:
* contributions never collide (1 thread/cell)
* contributions are sorted by column index already
(correct me if any of this is wrong though)

Reply:
To be clear, this only applies to cell-centered (flow) equation rows of course

Reply:
Yes, that was clear. Your suggestion is definitely correct. The original motivation for the current implementation is that I wanted to have a coupled solver that could also handle a pure finite-element formulation, i.e. using nodal interpolation also for the pressure field.

Reply:
The previous comment applies exclusively to single-phase poromechanics. I assume that for multiphase flow we will only support cell-centered (possibly hybridized) discretizations. @rrsettgast, @joshua-white would you agree?

Reply:
Actually, now that I've read it again, `localRowDofIndex` and `localFlowDispJacobian` are not necessarily sorted? So the first one would be `addToRowBinarySearchUnsorted< serialAtomic >`. Sorry for the confusion.

Reply:
Yeah, IMO I am not even sure we will have a purely FEM flow solver any time soon. If we wanted to do it for multiphase we would also have to use a mix formulation or another way to enforce mass conservation so I don't know if it's worth worrying about it for now.

Reply:
Done